File: experiments/pm-synth/src/audio/modules/interface/strategies/BrassStrategy.js
// BrassStrategy.js
// Brass interface: Asymmetric lip model with different positive/negative slopes

import { InterfaceStrategy } from '../InterfaceStrategy.js';
import { fastTanh } from '../utils/NonlinearityLib.js';

export class BrassStrategy extends InterfaceStrategy {
    constructor(sampleRate) {
        super(sampleRate);
    }

    /**
     * Process one sample through brass interface
     * Asymmetric lip buzz nonlinearity
     * @param {number} input - Input sample
     * @returns {number} Processed output
     */
    process(input) {
        const drive = 1.5 + this.intensity * 5;
        let shaped;

        if (input >= 0) {
            const lifted = input * drive + (0.2 + this.intensity * 0.35);
            shaped = fastTanh(Math.max(lifted, 0));
        } else {
            const compressed = -input * (drive * (0.4 + this.intensity * 0.4));
            shaped = -Math.pow(Math.min(compressed, 1.5), 1.3) * (0.35 + (1 - this.intensity) * 0.25);
        }

        const buzz = fastTanh(shaped * (1.2 + this.intensity * 1.5));
        return Math.max(-1, Math.min(1, buzz + this.intensity * 0.05));
    }

    /**
     * Reset state (stateless)
     */
    reset() {
        // No state to reset
    }
}

File: flues-synth/src/audio/modules/strategies/brass_strategy.c
// brass_strategy.c
// Brass interface: Asymmetric lip model with different positive/negative slopes
// Translated from reference/modules/interface/strategies/BrassStrategy.js

#include "interface_strategy.h"
#include "dsp_utils.h"
#include <stdlib.h>
#include <math.h>

// BrassStrategy is stateless, so no impl_data needed

static float brass_process(InterfaceStrategy* self, float input) {
    // Reduced drive to prevent excessive output
    const float drive = 0.8f + self->intensity * 2.0f;
    float shaped;

    if (input >= 0.0f) {
        // Removed DC lift - asymmetry from processing difference only
        const float driven = input * drive;
        shaped = fast_tanh(driven);
    } else {
        // Asymmetric negative processing (compressed, shaped)
        const float compressed = -input * (drive * (0.4f + self->intensity * 0.4f));
        shaped = -powf(fminf(compressed, 1.5f), 1.3f) * (0.35f + (1.0f - self->intensity) * 0.25f);
    }

    // Reduced final drive and removed DC offset addition
    const float buzz = fast_tanh(shaped * (0.8f + self->intensity * 0.8f));
    // Apply DC correction to compensate for asymmetric processing
    const float output = buzz * 0.35f - 0.02f;  // Small DC offset from asymmetry
    return fmaxf(-1.0f, fminf(1.0f, output));
}

static void brass_reset(InterfaceStrategy* self) {
    // Stateless, nothing to reset
    (void)self;
}

static void brass_set_intensity(InterfaceStrategy* self, float intensity) {
    self->intensity = intensity;
}

static void brass_set_gate(InterfaceStrategy* self, bool gate) {
    // No gate behavior for brass
    (void)self;
    (void)gate;
}

static void brass_destroy(InterfaceStrategy* self) {
    if (self) {
        free(self);
    }
}

static InterfaceStrategyVTable brass_vtable = {
    .process = brass_process,
    .reset = brass_reset,
    .set_intensity = brass_set_intensity,
    .set_gate = brass_set_gate,
    .destroy = brass_destroy
};

InterfaceStrategy* brass_strategy_create(float sample_rate) {
    (void)sample_rate;  // Unused
    InterfaceStrategy* strategy = (InterfaceStrategy*)calloc(1, sizeof(InterfaceStrategy));
    if (!strategy) return NULL;

    strategy->vtable = &brass_vtable;
    strategy->impl_data = NULL;  // Stateless
    strategy->intensity = 0.5f;

    return strategy;
}
